File: web/static/modules/stream-contract.js
export const STREAM_EVENT_TYPES = {
  HEARTBEAT: 'heartbeat',
  CONTENT: 'content',
  REASONING: 'reasoning',
  TOOL_CALL: 'tool_call',
  ERROR: 'error',
  MEMORY: 'memory'
};

export const STREAM_EVENT_VALUES = Object.values(STREAM_EVENT_TYPES);

export function isStreamEventType(type) {
  return STREAM_EVENT_VALUES.indexOf(type) >= 0;
}

export function parseStreamEvent(raw) {
  if (!raw) return null;
  var msg = raw;
  if (typeof raw === 'string') {
    try {
      msg = JSON.parse(raw);
    } catch (e) {
      return null;
    }
  }
  if (!msg || typeof msg !== 'object') return null;
  if (typeof msg.t !== 'string' || !isStreamEventType(msg.t)) return null;
  return { t: msg.t, d: msg.d };
}

export function serializeStreamEvent(type, data) {
  if (!isStreamEventType(type)) {
    throw new Error('Unsupported stream event type: ' + type);
  }
  return JSON.stringify({ t: type, d: data }) + '\n';
}

File: web/services/stream_contract.py
"""Shared contract for NDJSON stream events."""

from __future__ import annotations

import json
from typing import Any

STREAM_EVENT_TYPES = frozenset({
    "heartbeat",
    "content",
    "reasoning",
    "tool_call",
    "error",
    "memory",
    "notification",
    "retry",
})


def build_stream_event(event_type: str, data: Any) -> dict[str, Any]:
    if event_type not in STREAM_EVENT_TYPES:
        raise ValueError(f"Unsupported stream event type: {event_type}")
    return {"t": event_type, "d": data}


def serialize_stream_event(event_type: str, data: Any) -> str:
    return json.dumps(build_stream_event(event_type, data), ensure_ascii=False) + "\n"
